File: DomaskaNaWeekend/listTwo.cpp
#include<iostream>
#include"listTwo.h"
using namespace std;

void addNodeHeadTwo(ForwardListTwo* flist, const string& num) {
    DoublyNode* newNode = new DoublyNode;
    newNode->node = num;
    newNode->next = flist->head;
    newNode->prev = nullptr;

    if (flist->head != nullptr) {
        flist->head->prev = newNode;
    }
    flist->head = newNode;

    if (flist->tail == nullptr) {
        flist->tail = newNode;
    }
}

void addNodeTailTwo(ForwardListTwo* flist, const string& num) {
    DoublyNode* newNode = new DoublyNode;
    newNode->node = num;
    newNode->next = nullptr;
    newNode->prev = flist->tail;

    if (flist->tail != nullptr) {
        flist->tail->next = newNode;
    }

    flist->tail = newNode;

    if (flist->head == nullptr) {
        flist->head = newNode;
    }
}

void addNodeAfterTwo(ForwardListTwo* flist, DoublyNode* target, const string& num) {
    if (target != nullptr && flist->head != nullptr) {
        DoublyNode* newNode = new DoublyNode;
        newNode->node = num;
        newNode->next = nullptr;
        newNode->prev = nullptr;

        // Проверяем, что target действительно находится в списке
        DoublyNode* current = flist->head;
        bool targetFound = false;

        while (current != nullptr) {
            if (current == target) {
                targetFound = true;
                break;
            }
            current = current->next;
        }

        if (targetFound) {
            // Вставляем после target
            newNode->next = target->next;
            newNode->prev = target;

            if (target->next != nullptr) {
                target->next->prev = newNode;
            }
            else {
                flist->tail = newNode;
            }
            target->next = newNode;
        }
        else {
            // target не принадлежит списку, освобождаем память
            delete newNode;
            cout << "Ошибка: указанный узел не найден в списке" << endl;
        }
    }
}

void addNodeBeforeTwo(ForwardListTwo* flist, DoublyNode* target, const string& num) {
    if (target != nullptr && flist->head != nullptr) {
        DoublyNode* newNode = new DoublyNode;
        newNode->node = num;
        newNode->next = nullptr;
        newNode->prev = nullptr;

        // Проверяем, что target действительно находится в списке
        DoublyNode* current = flist->head;
        bool targetFound = false;

        while (current != nullptr) {
            if (current == target) {
                targetFound = true;
                break;
            }
            current = current->next;
        }

        if (targetFound) {
            // Вставляем перед target
            newNode->next = target;
            newNode->prev = target->prev;

            if (target->prev != nullptr) {
                target->prev->next = newNode;
            }
            else {
                flist->head = newNode;
            }
            target->prev = newNode;
        }
        else {
            // target не принадлежит списку, освобождаем память
            delete newNode;
            cout << "Ошибка: указанный узел не найден в списке" << endl;
        }
    }
}



void deleteNodeHeadTwo(ForwardListTwo* flist) {
    if (flist->head != nullptr) {
        DoublyNode* toDelete = flist->head;
        flist->head = flist->head->next;
        if (flist->head != nullptr) {
            flist->head->prev = nullptr;
        }
        else {
            flist->tail = nullptr;
        }
        delete toDelete;
    }
}

void deleteNodeTailTwo(ForwardListTwo* flist) {
    if (flist->tail != nullptr) {
        DoublyNode* toDelete = flist->tail;
        flist->tail = flist->tail->prev;
        if (flist->tail != nullptr) {
            flist->tail->next = nullptr;
        }
        else {
            flist->head = nullptr;
        }
        delete toDelete;
    }
}

void deleteNodeAfterTwo(ForwardListTwo* flist, DoublyNode* target) {
    if (target != nullptr && flist->head != nullptr) {
        // Проверяем, что target действительно находится в списке
        DoublyNode* current = flist->head;
        bool targetFound = false;

        while (current != nullptr) {
            if (current == target) {
                targetFound = true;
                break;
            }
            current = current->next;
        }

        if (targetFound && target->next != nullptr) {
            DoublyNode* toDelete = target->next;
            target->next = toDelete->next;

            if (toDelete->next != nullptr) {
                toDelete->next->prev = target;
            }
            else {
                flist->tail = target;
            }
            delete toDelete;
        }
    }
}

void deleteNodeBeforeTwo(ForwardListTwo* flist, DoublyNode* target) {
    if (target != nullptr && flist->head != nullptr) {
        // Проверяем, что target действительно находится в списке
        DoublyNode* current = flist->head;
        bool targetFound = false;

        while (current != nullptr) {
            if (current == target) {
                targetFound = true;
                break;
            }
            current = current->next;
        }

        if (targetFound && target->prev != nullptr) {
            DoublyNode* toDelete = target->prev;
            target->prev = toDelete->prev;

            if (toDelete->prev != nullptr) {
                toDelete->prev->next = target;
            }
            else {
                flist->head = target;
            }
            delete toDelete;
        }
    }
}

bool deleteNodTwo(ForwardListTwo* flist, const string& num) {
    if (flist->head == nullptr) return false;

    DoublyNode* current = flist->head;

    while (current != nullptr && current->node != num) {
        current = current->next;
    }

    if (current == nullptr) return false;

    if (current->prev != nullptr) {
        current->prev->next = current->next;
    }
    else {
        flist->head = current->next;
    }

    if (current->next != nullptr) {
        current->next->prev = current->prev;
    }
    else {
        flist->tail = current->prev;
    }

    delete current;
    return true;
}

bool findNodeIndexTwo(ForwardListTwo* flist, const string& num) {
    DoublyNode* current = flist->head;
    while (current != nullptr) {
        if (current->node == num) return true;
        current = current->next;
    }
    return false;
}

void printListTwo(const ForwardListTwo& flist) {
    DoublyNode* current = flist.head;
    cout << "Список: ";
    while (current != nullptr) {
        cout << current->node << " ";
        current = current->next;
    }
    cout << endl;
}

int countNodesTwo(const ForwardListTwo& flist) {
    int count = 0;
    DoublyNode* current = flist.head;
    while (current != nullptr) {
        count++;
        current = current->next;
    }
    return count;
}

DoublyNode* getNodeByIndexTwo(ForwardListTwo& flist, int index) {
    if (index < 0) return nullptr;

    DoublyNode* current = flist.head;
    int currentIndex = 0;

    while (current != nullptr) {
        if (currentIndex == index) return current;
        current = current->next;
        currentIndex++;
    }

    return nullptr;
}

bool deleteNodeIndexTwo(ForwardListTwo* flist, const string& num) {
    if (flist->head == nullptr) return false;

    DoublyNode* current = flist->head;

    while (current != nullptr && current->node != num) {
        current = current->next;
    }

    if (current == nullptr) return false;

    if (current->prev != nullptr) {
        current->prev->next = current->next;
    }
    else {
        flist->head = current->next;
    }

    if (current->next != nullptr) {
        current->next->prev = current->prev;
    }
    else {
        flist->tail = current->prev;
    }

    delete current;
    return true;
}

File: Godomaska/list_two.go
package main

import "fmt"

type DoublyNode struct {
	node string
	next *DoublyNode
	prev *DoublyNode
}

type ForwardListTwo struct {
	head *DoublyNode
	tail *DoublyNode
}

func NewDoublyList() *ForwardListTwo {
	return &ForwardListTwo{head: nil, tail: nil}
}

func addNodeHeadTwo(flist *ForwardListTwo, num string) {
	newNode := &DoublyNode{node: num, next: flist.head, prev: nil}
	if flist.head != nil {
		flist.head.prev = newNode
	}
	flist.head = newNode
	if flist.tail == nil {
		flist.tail = newNode
	}
}

func addNodeTailTwo(flist *ForwardListTwo, num string) {
	newNode := &DoublyNode{node: num, next: nil, prev: flist.tail}
	if flist.tail != nil {
		flist.tail.next = newNode
	}
	flist.tail = newNode
	if flist.head == nil {
		flist.head = newNode
	}
}

func addNodeAfterTwo(flist *ForwardListTwo, target *DoublyNode, num string) {
	if target != nil {
		newNode := &DoublyNode{node: num, next: target.next, prev: target}
		if target.next != nil {
			target.next.prev = newNode
		} else {
			flist.tail = newNode
		}
		target.next = newNode
	}
}

func addNodeBeforeTwo(flist *ForwardListTwo, target *DoublyNode, num string) {
	if target != nil {
		newNode := &DoublyNode{node: num, next: target, prev: target.prev}
		if target.prev != nil {
			target.prev.next = newNode
		} else {
			flist.head = newNode
		}
		target.prev = newNode
	}
}

func deleteNodeHeadTwo(flist *ForwardListTwo) {
	if flist.head != nil {
		toDelete := flist.head
		flist.head = flist.head.next
		if flist.head != nil {
			flist.head.prev = nil
		} else {
			flist.tail = nil
		}
		toDelete.next = nil
		toDelete.prev = nil
	}
}

func deleteNodeTailTwo(flist *ForwardListTwo) {
	if flist.tail != nil {
		toDelete := flist.tail
		flist.tail = flist.tail.prev
		if flist.tail != nil {
			flist.tail.next = nil
		} else {
			flist.head = nil
		}
		toDelete.next = nil
		toDelete.prev = nil
	}
}

func deleteNodeAfterTwo(flist *ForwardListTwo, target *DoublyNode) {
	if target != nil && target.next != nil {
		toDelete := target.next
		target.next = toDelete.next
		if toDelete.next != nil {
			toDelete.next.prev = target
		} else {
			flist.tail = target
		}
		toDelete.next = nil
		toDelete.prev = nil
	}
}

func deleteNodeBeforeTwo(flist *ForwardListTwo, target *DoublyNode) {
	if target != nil && target.prev != nil {
		toDelete := target.prev
		target.prev = toDelete.prev
		if toDelete.prev != nil {
			toDelete.prev.next = target
		} else {
			flist.head = target
		}
		toDelete.next = nil
		toDelete.prev = nil
	}
}

func deleteNodeIndexTwo(flist *ForwardListTwo, num string) bool {
	if flist.head == nil {
		return false
	}
	cur := flist.head
	for cur != nil {
		if cur.node == num {
			if cur.prev != nil {
				cur.prev.next = cur.next
			} else {
				flist.head = cur.next
			}
			if cur.next != nil {
				cur.next.prev = cur.prev
			} else {
				flist.tail = cur.prev
			}
			return true
		}
		cur = cur.next
	}
	return false
}

func findNodeIndexTwo(flist *ForwardListTwo, num string) bool {
	cur := flist.head
	for cur != nil {
		if cur.node == num {
			return true
		}
		cur = cur.next
	}
	return false
}

func printListTwo(flist ForwardListTwo) {
	fmt.Print("Список: ")
	cur := flist.head
	for cur != nil {
		fmt.Print(cur.node, " ")
		cur = cur.next
	}
	fmt.Println()
}

func countNodesTwo(flist ForwardListTwo) int {
	cnt := 0
	cur := flist.head
	for cur != nil {
		cnt++
		cur = cur.next
	}
	return cnt
}

func getNodeByIndexTwo(flist ForwardListTwo, index int) *DoublyNode {
	if index < 0 {
		return nil
	}
	cur := flist.head
	i := 0
	for cur != nil {
		if i == index {
			return cur
		}
		cur = cur.next
		i++
	}
	return nil
}
